website/src/app/blog/sans-io/readme.mdx

```diff
@@ -0,0 +1,693 @@
+errors simply becomes part of state machine's input handling which leads to more
+robust code and makes it less likely for edge-cases to only be considered as an
+after-thought.
+
+## Reflecting on Rust's features
+
+With an understanding of sans-IO, let's focus our attention on a few of Rust's
+features:
+
+- ownership
+- explicit, mutable borrowing
+- exhaustive matching
+- powerful algebraic data types, especially enums
+
+### Ownership & mutable borrowing
+
+Rust forces us to declare, which component or function in our code owns a
+certain value. A common example for these are buffers: When reading from a
+`UdpSocket`, we need to provide a `&mut [u8]` as a place for the actual bytes
+being received. Only the owner of a value can declare it mutable and thus either
+mutate itself or temporarily hand out mutable references to other functions.
+`UdpSocket` follows this design: It doesn't declare a buffer on its own,
+instead, it only requires temporary, mutable access to it when it is actually
+reading from the socket.
+
+### Exhaustive matching and ADTs
+
+The `match` keyword provides us with a compile-time checked way of handling the
+different variations of a value. Combined with enums, this provides a powerful
+way of ensuring that all combinations are handled. Enums lend themselves
+especially well in modelling state that can change at runtime. The following
+code is a slightly simplified snippet from
+[actual production code](https://github.com/firezone/firezone/blob/79fd8f6063761e582013873ff2a49d22c2445c46/rust/connlib/snownet/src/node.rs#L1308-L1330),
+modelling the state of a connection:
+
+```rust
+enum Connection {
+    /// We are still running ICE to figure out, which socket to use to send data.
+    Connecting {
+        /// Socket addresses from which we might receive data (even before we are connected).
+        possible_sockets: HashSet<SocketAddr>,
+        /// Packets emitted by wireguard whilst are still running ICE.
+        ///
+        /// This can happen if the remote's WG session initiation arrives at our socket before we nominate it.
+        /// A session initiation requires a response that we must not drop, otherwise the connection setup experiences unnecessary delays.
+        buffered: RingBuffer<Vec<u8>>,
+    },
+    /// A socket has been nominated.
+    Connected {
+        /// Our nominated socket.
+        peer_socket: PeerSocket,
+        /// Other addresses that we might see traffic from (e.g.
+        STUN messages during roaming).
+        possible_sockets: HashSet<SocketAddr>,
+    },
+    /// The connection failed in an unrecoverable way and will be GC'd.
+    Failed,
+    /// The connection is idle and will be GC'd.
+    Idle,
+}
+```
+
+The key things about the above model are the invariants around which data is
+available in a particular state. For example, buffered packets can only happen
+whilst we are Connecting. Once we are Connected, there is no more need to buffer
```

Comment:
For me, the importance of `&mut` can't be overstated. To me that's even more important than function coloring. To elaborate:

Maybe the most central concept in Rust, or the reason for it to exist as a language, is the borrow checker: which is all about the safe use of `&mut`.

In async code it's rarely appropriate to hold `&mut` references across an `.await` point to anything which isn't local to the task.

Thus in some respects async is at odds with the most fundamental part of Rust. 

Arc-Mutex, actors/channels, interior mutability all have their place, but async forces us to reach for them straight away. I think this often results in less good design and spaghetti code. 

Sans-IO allows us to structure the code in a lock free way by passing around `&mut` references – and it really brings out how great Rust is (sync Rust that is).

Reply:
Hm. You do bring out these points below. I'll leave my comment anyway :)

Reply:
Do you think it comes out early enough? I could perhaps trim this section and remove the `Connection` bit. Most people probably don't need to be reminded of Rust's ADT features. That would place more emphasis on `&mut`.

I'll try this and see how it feels.